commit 90fcc606cd55456d510bc228916eb7a2b511ca12 (refs/changes/91/891/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2016-11-14T12:14:15+01:00 (2 years, 11 months ago)
    
    CNAPI-677 Add support for disklayout params to server setup FIXED
    Reviewed by: Orlando Vazquez <orlando@joyent.com>
    Reviewed by: Trent Mick <trent.mick@joyent.com>
    
    This reverts commit 7034ad8ebfa08cfaacf0bd914580d2173f985e7d.

diff --git a/docs/index.md b/docs/index.md
index 901774c..90bd708 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -1153,6 +1153,9 @@ Initiate the server setup process for a newly started server.
 | nics             | Object | Nic parameters to update                 |
 | postsetup_script | String | Script to run after setup has completed  |
 | hostname         | String | Hostname to set for the specified server |
+| disk_layout      | String | Disk layout type (man disklayout)        |
+| disk_spares      | Number | Number of disk spares (man disklayout)   |
+| disk_cache       | String | Disk cache ('true'|'false')              |
 
 
 ### Responses
diff --git a/lib/endpoints/servers.js b/lib/endpoints/servers.js
index 34d2b95..ef81a3c 100644
--- a/lib/endpoints/servers.js
+++ b/lib/endpoints/servers.js
@@ -1,5 +1,14 @@
 /*
- * Copyright 2016, Joyent, Inc. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2016, Joyent, Inc.
+ */
+
+/*
  *
  * HTTP endpoints for interacting with compute nodes.
  *
@@ -604,6 +613,10 @@ Server.factoryReset = function (req, res, next) {
  * @param {Object} nics Nic parameters to update
  * @param {String} postsetup_script Script to run after setup has completed
  * @param {String} hostname Hostname to set for the specified server
+ * @param {String} disk_spares See `man disklayout` spares
+ * @param {String} disk_cache See `man disklayout` cache
+ * @param {String} disk_layout See `man disklayout` type
+ *      (single, mirror, raidz1, ...)
  * @response 200 Object Setup initated, returns object containing workflow id
  * @response 500 None Error while processing request
  */
@@ -614,7 +627,10 @@ Server.setup = function (req, res, next) {
         'hostname': ['optional', 'isStringType'],
         'nics': ['optional', 'isArrayType'],
         'postsetup_script': ['optional', 'isStringType'],
-        'server_uuid': ['isStringType']
+        'server_uuid': ['isStringType'],
+        'disk_spares': ['optional', 'isNumberGreaterThanEqualZeroType'],
+        'disk_cache': ['optional', 'isBooleanString'],
+        'disk_layout': ['optional', 'isStringType']
     };
 
     if (validation.ensureParamsValid(req, res, rules, { strict: true })) {
@@ -639,6 +655,26 @@ Server.setup = function (req, res, next) {
     if (req.params.hasOwnProperty('creator_uuid')) {
         params.creator_uuid = req.params.creator_uuid;
     }
+    if (typeof (req.params.disk_spares) !== 'undefined') {
+        params.disk_spares = req.params.disk_spares;
+    }
+    if (typeof (req.params.disk_cache) !== 'undefined') {
+        params.disk_cache = (req.params.disk_cache === 'true');
+    }
+    if (typeof (req.params.disk_layout) !== 'undefined') {
+        var layout = req.params.disk_layout;
+        var VALID_LAYOUTS = ['single', 'mirror', 'raidz1', 'raidz2', 'raidz3'];
+        if (VALID_LAYOUTS.indexOf(layout) === -1) {
+            var err = new restify.InvalidArgumentError(
+                sprintf('disk_layout must be one of: \'%s\'',
+                    VALID_LAYOUTS.join('\', \'')));
+            res.send(err);
+            next();
+            return;
+        }
+        params.disk_layout = layout;
+    }
+
     req.stash.server.getRaw(function (error, rawserver) {
         if (rawserver.setup) {
             res.send(204);
diff --git a/lib/models/server.js b/lib/models/server.js
index 2264fbf..334d75e 100644
--- a/lib/models/server.js
+++ b/lib/models/server.js
@@ -1584,6 +1584,21 @@ ModelServer.prototype.setup = function (params, callback) {
             wfParams.creator_uuid = params.creator_uuid;
         }
 
+        if (params.hasOwnProperty('disk_spares')) {
+            wfParams.disk_spares = params.disk_spares;
+        }
+
+        // Caching is the default, we only need to pass in disk cache
+        // when it's false:
+        if (params.hasOwnProperty('disk_cache') &&
+            params.disk_cache === false) {
+            wfParams.disk_cache = params.disk_cache;
+        }
+
+        if (params.hasOwnProperty('disk_layout') && params.disk_layout) {
+            wfParams.disk_layout = params.disk_layout;
+        }
+
         async.waterfall([
             function (cb) {
                 if (params.hasOwnProperty('hostname') && params.hostname) {
diff --git a/lib/workflows/server-setup.js b/lib/workflows/server-setup.js
index 92dce66..25c4f2a 100644
--- a/lib/workflows/server-setup.js
+++ b/lib/workflows/server-setup.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2016, Joyent, Inc.
  */
 
 /*
@@ -24,7 +24,7 @@
  * - Set server setup => true
  */
 
-var VERSION = '1.0.2';
+var VERSION = '1.0.3';
 
 var cnapiCommon = require('wf-shared').cnapi;
 var napiCommon = require('wf-shared').napi;
@@ -82,6 +82,16 @@ function fetchSetupFiles(job, callback) {
     var cnapi = restify.createJsonClient({ url: cnapiUrl});
     var overprovision_ratio = job.params.overprovision_ratio || '1.0';
     var nodeConfigDir = 'config-' + job.params.server_uuid;
+    var diskCfg = '';
+    if (job.params.disk_layout) {
+        diskCfg += 'echo "layout=\'' + job.params.disk_layout + '\'"; ';
+    }
+    if (job.params.hasOwnProperty('disk_cache')) {
+        diskCfg += 'echo "cache=\'' + job.params.disk_cache + '\'"; ';
+    }
+    if (job.params.hasOwnProperty('disk_spares')) {
+        diskCfg += 'echo "spares=' + job.params.disk_spares + '"; ';
+    }
 
     var script = [
         '#!/bin/bash',
@@ -89,9 +99,10 @@ function fetchSetupFiles(job, callback) {
         'cd /var/tmp',
         'mkdir /var/tmp/node.config',
         'mkdir /var/tmp/' + nodeConfigDir,
-        '(echo "overprovision_ratio=\'' + overprovision_ratio + '\'"; '
-            + 'curl $1/extra/joysetup/node.config) | tee ' + nodeConfigDir +
-              '/node.config node.config/node.config >&-',
+        '(echo "overprovision_ratio=\'' + overprovision_ratio + '\'"; ' +
+             diskCfg +
+             'curl $1/extra/joysetup/node.config) | tee ' + nodeConfigDir +
+             '/node.config node.config/node.config >&-',
         'curl -O $1/extra/joysetup/joysetup.sh',
         'curl -O $1/extra/joysetup/agentsetup.sh',
         'chmod +x *.sh'
